File: cli/src/verdicts.rs
//! Streaming JSON Lines output of every reference verdict (`check --verdicts-json`).
//!
//! Evaluation surface for tools/jvm-probe: each line carries the raw reference as
//! extracted from the constant pool (never collapsed the way report violations
//! are) plus the verdict it received. Lines are streamed as verdicts are
//! computed, so memory stays flat regardless of scan size. Exclude rules do not
//! apply here; they filter the report, not the verdict stream.

use crate::intern::Sym;
use crate::model::SymbolRef;
use anyhow::{Context, Result};
use serde::Serialize;
use std::io::{BufWriter, Write};
use std::path::Path;

#[derive(Serialize)]
struct VerdictRecord<'a> {
    source: Sym,
    source_class: Sym,
    reference: &'a SymbolRef,
    verdict: &'static str,
    #[serde(skip_serializing_if = "Option::is_none")]
    reason: Option<&'static str>,
}

/// Writes one JSON line per reference verdict. A write failure is remembered and
/// surfaced once by `finish` instead of aborting the check mid-scan.
pub struct VerdictWriter {
    out: Option<Box<dyn Write>>,
    error: Option<String>,
}

impl VerdictWriter {
    pub fn create(path: &Path) -> Result<Self> {
        let file = std::fs::File::create(path)
            .with_context(|| format!("cannot create verdicts output {}", path.display()))?;
        Ok(Self {
            out: Some(Box::new(BufWriter::new(file))),
            error: None,
        })
    }

    pub fn record(
        &mut self,
        source: Sym,
        source_class: Sym,
        reference: &SymbolRef,
        verdict: &'static str,
        reason: Option<&'static str>,
    ) {
        let Some(out) = self.out.as_mut() else { return };
        let rec = VerdictRecord {
            source,
            source_class,
            reference,
            verdict,
            reason,
        };
        let result = (|| -> std::io::Result<()> {
            serde_json::to_writer(&mut *out, &rec)?;
            out.write_all(b"\n")
        })();
        if let Err(e) = result {
            self.error = Some(format!("verdicts output failed, stream truncated: {e}"));
            self.out = None;
        }
    }

    /// Flush and return the deferred write error, if any.
    pub fn finish(mut self) -> Option<String> {
        if let Some(out) = self.out.as_mut()
            && let Err(e) = out.flush()
        {
            return Some(format!("verdicts output failed, stream truncated: {e}"));
        }
        self.error.take()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::intern::intern;
    use crate::model::{MemberKey, RefKind};

    #[test]
    fn streams_one_json_line_per_record() {
        let dir = std::env::temp_dir().join(format!("uika-verdicts-test-{}", std::process::id()));
        std::fs::create_dir_all(&dir).unwrap();
        let path = dir.join("verdicts.jsonl");

        let mut w = VerdictWriter::create(&path).unwrap();
        let broken = SymbolRef {
            kind: RefKind::Method,
            owner: intern("com/example/Owner"),
            member: Some(MemberKey::new("gone", "()V")),
            expected_static: Some(true),
            field_write: None,
        };
        w.record(
            intern("app.jar"),
            intern("com/example/Caller"),
            &broken,
            "broken",
            Some("method removed"),
        );
        let ok = SymbolRef {
            kind: RefKind::Class,
            owner: intern("com/example/Owner"),
            member: None,
            expected_static: None,
            field_write: None,
        };
        w.record(
            intern("app.jar"),
            intern("com/example/Caller"),
            &ok,
            "ok",
            None,
        );
        assert!(w.finish().is_none());

        let text = std::fs::read_to_string(&path).unwrap();
        let lines: Vec<&str> = text.lines().collect();
        assert_eq!(lines.len(), 2);
        let first: serde_json::Value = serde_json::from_str(lines[0]).unwrap();
        assert_eq!(first["verdict"], "broken");
        assert_eq!(first["reason"], "method removed");
        assert_eq!(first["reference"]["owner"], "com/example/Owner");
        assert_eq!(first["reference"]["member"]["name"], "gone");
        assert_eq!(first["reference"]["expected_static"], true);
        let second: serde_json::Value = serde_json::from_str(lines[1]).unwrap();
        assert_eq!(second["verdict"], "ok");
        assert_eq!(second["reference"]["kind"], "class");
        assert!(second.get("reason").is_none());

        let _ = std::fs::remove_dir_all(&dir);
    }
}

File: cli-java/src/main/java/net/exoego/uika/cli/Verdicts.java
package net.exoego.uika.cli;

import java.io.BufferedOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * Streaming JSON Lines output of every reference verdict ({@code --verdicts-json}).
 *
 * <p>The evaluation surface for tools/jvm-probe: each line carries the raw reference as it
 * sits in the constant pool plus the verdict it received, one line per reference record with
 * no call-site dedup. Lines are streamed as verdicts are computed, so memory stays flat.
 */
final class Verdicts {
    private Verdicts() {}

    static final class Writer {
        /** Null once the stream has failed. */
        private OutputStream out;
        /** The first failure, when the stream has failed. */
        private String failure;
        private String module;
        private final StringBuilder line = new StringBuilder(256);

        private Writer(OutputStream out) {
            this.out = out;
        }

        static Writer create(String path) {
            try {
                return new Writer(new BufferedOutputStream(Files.newOutputStream(Path.of(path)), 1 << 16));
            } catch (IOException e) {
                throw new UikaException("cannot create verdicts output " + path, e);
            }
        }

        static Writer to(OutputStream out) {
            return new Writer(out);
        }

        /** Labels subsequent records with the per-module run producing them; null clears it. */
        void setModule(String module) {
            this.module = module;
        }

        void record(int source, int sourceClass, SymbolRef reference, String verdict, String reason) {
            if (out == null) {
                return;
            }
            line.setLength(0);
            Json.Writer json = new Json.Writer(line, false);
            json.beginObject();
            json.key("source").sym(source);
            json.key("source_class").sym(sourceClass);
            json.key("reference");
            writeReference(json, reference);
            json.key("verdict").value(verdict);
            if (reason != null) {
                json.key("reason").value(reason);
            }
            if (module != null) {
                json.key("module").value(module);
            }
            json.endObject();
            line.append('\n');
            try {
                out.write(line.toString().getBytes(StandardCharsets.UTF_8));
            } catch (IOException e) {
                // Latched: the scan goes on, but the command fails afterwards. A truncated
                // stream must never pass silently, the probe would answer-check only a prefix.
                failure = truncated(e);
                closeQuietly();
            }
        }

        /** Flushes, and returns the failure message if the stream failed at any point. */
        String finish() {
            if (out == null) {
                return failure;
            }
            try {
                out.flush();
                out.close();
                out = null;
                return null;
            } catch (IOException e) {
                out = null;
                return truncated(e);
            }
        }

        private void closeQuietly() {
            try {
                out.close();
            } catch (IOException ignored) {
                // the first failure is already latched
            }
            out = null;
        }

        private static String truncated(IOException e) {
            return "verdicts output failed, stream truncated: " + UikaException.describe(e);
        }
    }

    /** The reference object as every JSON surface spells it. */
    static void writeReference(Json.Writer json, SymbolRef reference) {
        json.beginObject();
        json.key("kind").value(reference.kind().json);
        json.key("owner").sym(reference.owner());
        if (reference.hasMember()) {
            json.key("member").beginObject();
            json.key("name").sym(MemberKey.name(reference.member()));
            json.key("descriptor").sym(MemberKey.descriptor(reference.member()));
            json.endObject();
        }
        if (reference.expectedStatic() != null) {
            json.key("expected_static").value(reference.expectedStatic());
        }
        if (reference.fieldWrite() != null) {
            json.key("field_write").value(reference.fieldWrite());
        }
        if (reference.instantiated() != null) {
            json.key("instantiated").value(reference.instantiated());
        }
        json.endObject();
    }
}
